src/OrchardCore/OrchardCore/Environment/Shell/RunningShellTable.cs

```diff
@@ -1,103 +1,80 @@
 using System;
-using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
-using System.Threading;
 
 namespace OrchardCore.Environment.Shell
 {
     public class RunningShellTable : IRunningShellTable
     {
-        private readonly Dictionary<string, ShellSettings> _shellsByHostAndPrefix = new Dictionary<string, ShellSettings>(StringComparer.OrdinalIgnoreCase);
-        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim();
+        private static readonly char[] HostSeparators = new[] { ',', ' ' };
 
+        private ImmutableDictionary<string, ShellSettings> _shellsByHostAndPrefix = ImmutableDictionary<string, ShellSettings>.Empty.WithComparers(StringComparer.OrdinalIgnoreCase);
         private ShellSettings _default;
         private bool _hasStarMapping = false;
 
         public void Add(ShellSettings settings)
         {
-            _lock.EnterWriteLock();
-            try
+            if (ShellHelper.DefaultShellName == settings.Name)
             {
-                if(ShellHelper.DefaultShellName == settings.Name)
-                {
-                    _default = settings;
-                }
-
-                var allHostsAndPrefix = GetAllHostsAndPrefix(settings);
-                foreach (var hostAndPrefix in allHostsAndPrefix)
-                {
-                    _hasStarMapping = _hasStarMapping || hostAndPrefix.StartsWith("*");
-                    _shellsByHostAndPrefix[hostAndPrefix] = settings;
-                }
+                _default = settings;
             }
-            finally
+
+            var allHostsAndPrefix = GetAllHostsAndPrefix(settings);
+
+            foreach (var hostAndPrefix in allHostsAndPrefix)
             {
-                _lock.ExitWriteLock();
+                _hasStarMapping = _hasStarMapping || hostAndPrefix.StartsWith("*");
+                _shellsByHostAndPrefix = _shellsByHostAndPrefix.SetItem(hostAndPrefix, settings);
             }
         }
 
         public void Remove(ShellSettings settings)
         {
-            _lock.EnterWriteLock();
-            try
-            {
-                var allHostsAndPrefix = GetAllHostsAndPrefix(settings);
-                foreach (var hostAndPrefix in allHostsAndPrefix)
-                {
-                    _shellsByHostAndPrefix.Remove(hostAndPrefix);
-                }
+            var allHostsAndPrefix = GetAllHostsAndPrefix(settings);
 
-                if (_default == settings)
-                {
-                    _default = null;
-                }
+            foreach (var hostAndPrefix in allHostsAndPrefix)
+            {
+                _shellsByHostAndPrefix = _shellsByHostAndPrefix.Remove(hostAndPrefix);
             }
```

Comment:
I think using `RemoveRange` is optimized

    _shellsByHostAndPrefix = _shellsByHostAndPrefix.RemoveRange(allHostsAndPrefix);